File: java/src/main/java/vsdk/toolkit/raycasting/bidirectionalRaytracing/LightListIterator.java
package vsdk.toolkit.raycasting.bidirectionalRaytracing;

import vsdk.toolkit.common.dataStructures.CircularListIterator;
import vsdk.toolkit.skin.Patch;

public class LightListIterator {
    private CircularListIterator<LightInfo> iterator;

    public LightListIterator(LightList list) {
        iterator = new CircularListIterator<>(list.entries());
    }

    public Patch First(LightList list) {
        iterator.init(list.entries());

        LightInfo li = iterator.nextOnSequence();
        if ( li != null ) {
            return li.light;
        } else {
            return null;
        }
    }

    public Patch Next() {
        LightInfo li = iterator.nextOnSequence();
        if ( li != null ) {
            return li.light;
        } else {
            return null;
        }
    }
}

File: typescript/src/main/vsdk/toolkit/raycasting/bidirectionalRaytracing/LightListIterator.ts
import { CircularListIterator } from "../../common/dataStructures/CircularListIterator";
import { Patch } from "../../skin/Patch";
import { LightInfo } from "./LightInfo";
import { LightList } from "./LightList";

export class LightListIterator {
  private iterator: CircularListIterator<LightInfo>;

  public constructor(list: LightList) {
    this.iterator = new CircularListIterator<LightInfo>(list.entries());
  }

  public First(list: LightList): Patch | null {
    this.iterator.init(list.entries());

    const li = this.iterator.nextOnSequence();
    if (li !== null) {
      return li.light;
    }
    return null;
  }

  public Next(): Patch | null {
    const li = this.iterator.nextOnSequence();
    if (li !== null) {
      return li.light;
    }
    return null;
  }
}
